### PR TITLE
perf: optimize MXFP4xBF16 & INT4xFP8 CUTLASS MoE backend for SM90

### PR DESCRIPTION
**Pull request by samuellees**
## Summary

Port [TensorRT-LLM PR #12451](https://github.com/NVIDIA/TensorRT-LLM/pull/12451) to FlashInfer's `cutlass_fused_moe` SM90 path. Adds an LDSM + interleaved-LUT weight-load pipeline for 4-bit weights × 16/8-bit activations, plus the two preprocessing helpers the new kernel layout requires.

### Kernel
- `mixed_input_utils.hpp` / `sm90_mma_array_tma_gmma_rs_warpspecialized_mixed_input_.hpp` — sync with TRT-LLM PR #12451 (LDSM path + FP4/INT4 → BF16 LUT converter).
- `moe_gemm_mixed_utils.{cu,h}` (new) — per-row CUDA kernels for FP4/INT4 byte interleave.
- `cutlass_heuristic.cpp` — for `has_w4afp8`, skip `CtaShape128x128x128B + COOPERATIVE` (register overflow on SM90) and pick COOP / PINGPONG per tile.
- `moe_gemm_tma_ws_mixed_input_launcher.inl` — `scheduler.max_swizzle_size = 2`, `raster_order = Heuristic`.

### Python
`flashinfer/fused_moe/core.py` exposes two helpers (re-exported by the package):
- `interleave_moe_weights_for_hopper_mixed_gemm(weight, quant_type)` — byte-level interleave for `"fp4"` / `"int4"` packed uint8 weights; delegates to the C++ kernel above.
- `interleave_moe_scales_for_hopper_mixed_gemm(scales, group_size=32)` — pure PyTorch reshape + permute matching TRT-LLM's `WFP4A16FusedMoEMethod.load_quant_scales`, factor = `128 // group_size`.

### Tests — inside `tests/moe/test_trtllm_cutlass_fused_moe.py` (18 new)
- `test_moe_bf16_mxfp4_hopper_correctness` (5 shapes, strict `assert_close` vs a GPU-side dequantized reference that only materialises active experts to stay under H200 memory at e=256).
- `test_moe_bf16_mxfp4_hopper_coverage` (5 shapes, percent-based ≥ 99.9%).
- `test_moe_bf16_mxfp4_hopper_activations` (3 SwiGLU variants).
- `test_moe_w4a8_hopper_correctness` (2 shapes × bf16/fp16) — envelope matches the upstream CI shape (h = inter = 512, e = 2); larger exceeds strict tolerance because of FP8 + INT4 accumulation noise, same as the existing `test_moe_w4a8`.
- `test_moe_w4a8_hopper_autotune` — smoke that `autotune(True)` doesn't break the W4A8 path.

All 18 green on H200 in 5.2 s cache-hot.

## Performance

H200 (SM90 / HBM3e), `hidden = 4096, intermediate = 2048, experts = 256, topk = 6`, bf16 output, MXFP4 weights. `cutlass_fused_moe` median over `bench_gpu_time`. Weight + scale interleave is a one-shot model-load step and is excluded from timing. `autotune` column runs one pass under `autotune(True)` to populate the tactic cache before timing.

| batch | main no-autotune | main autotune | **PR no-autotune** | **PR autotune** | **speedup (autotune)** |
|------:|-----------------:|--------------:|-------------------:|----------------:|-----------------------:|
|   4   | 0.791 ms         | 0.513 ms      | **0.221 ms**       | **0.193 ms**    | **2.66×** |
|  16   | 1.598 ms         | 1.607 ms      | **0.530 ms**       | **0.532 ms**    | **3.02×** |
|  64   | 3.761 ms         | 3.757 ms      | **1.200 ms**       | **1.207 ms**    | **3.11×** |

### CODE REVIEW

File: tests/moe/test_trtllm_cutlass_fused_moe.py [hunk 1]
```diff
@@ -325,15 +325,15 @@ def compute_with_experts(
 
 # Test configurations
 BATCH_SIZES = [
-    1,
+    16,
```
Comment (jiahanc) [inline tests/moe/test_trtllm_cutlass_fused_moe.py]: can we keep both for better coverage?

**Review by jiahanc — APPROVED** [review]
LGTM. Leave 1 small comment, thanks for contribution!

File: csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu [hunk 2]
```diff
@@ -1286,4 +1275,48 @@ tvm::ffi::Module init(DLDataType activation_dtype, DLDataType weight_dtype, DLDa
   return tvm::ffi::Module(ptr);
 }
 
+// Interleave a 4-bit packed weight tensor into the layout required by the
+// SM90 mixed-input Hopper MoE GEMM. Expected input shape (num_experts, n,
+// k / 2) uint8 on CUDA. Writes into an output tensor of the same shape.
+// quant_type: 0 for INT4 (W4A8), 1 for FP4 (W4A16 / MXFP4).
+void interleave_moe_weights_for_Hopper_mixed_gemm(TensorView weight, TensorView weight_interleaved,
```
Comment (samuellees) [inline csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu]: repleace all Hopper/hopper with sm90 (function name)

File: csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu [hunk 3]
```diff
@@ -1286,4 +1275,48 @@ tvm::ffi::Module init(DLDataType activation_dtype, DLDataType weight_dtype, DLDa
   return tvm::ffi::Module(ptr);
 }
 
+// Interleave a 4-bit packed weight tensor into the layout required by the
+// SM90 mixed-input Hopper MoE GEMM. Expected input shape (num_experts, n,
```
Comment (samuellees) [inline csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu]: repleace Hopper sm90

File: csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu [hunk 4]
```diff
@@ -1286,4 +1275,48 @@ tvm::ffi::Module init(DLDataType activation_dtype, DLDataType weight_dtype, DLDa
...
+  int64_t const per_expert_bytes = n * (k / 2);
+
+  auto stream = get_stream(weight.device());
+  auto* src = static_cast<uint8_t*>(weight.data_ptr());
+  auto* dst = static_cast<uint8_t*>(weight_interleaved.data_ptr());
+  for (int64_t e = 0; e < num_experts; ++e) {
+    uint8_t* src_e = src + e * per_expert_bytes;
+    uint8_t* dst_e = dst + e * per_expert_bytes;
+    if (quant_type == 1) {
+      tensorrt_llm::kernels::cutlass_kernels::interleave_fp4_weights_for_Hopper_mixed_gemm(
```
Comment (samuellees) [inline csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu]: dito

File: csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/moe_gemm/moe_gemm_mixed_utils.cu [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+#include "moe_gemm_mixed_utils.h"
+
+namespace tensorrt_llm {
+namespace kernels {
+namespace cutlass_kernels {
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+__global__ void interleave_fp4_weights_for_Hopper_mixed_gemm_kernel(uint8_t* fp4_weight,
```
Comment (samuellees) [inline csrc/nv_internal/tensorrt_llm/kernels/cutlass_kernels/moe_gemm/moe_gemm_mixed_utils.cu]: dito

File: tests/moe/test_trtllm_cutlass_fused_moe.py [hunk 6]
```diff
@@ -2400,5 +2404,413 @@ def test_moe_mxfp8_mxfp4_ndim_padding_safety(
...
 
+# ============================================================================
+# SM90 mixed-input (Hopper) MoE tests — PR #3084
+#
+# Exercise the W4A16 (MXFP4 x BF16) and W4A8 (INT4 x FP8) paths with the
+# preprocessing helpers exposed by this PR: weights go through
+# ``interleave_moe_weights_for_hopper_mixed_gemm``, MXFP4 block scales go
+# through ``interleave_moe_scales_for_hopper_mixed_gemm``, and W4A8 weight
+# scales go through the group-size-4/2/1 reshape+permute already used by
+# ``test_moe_w4a8`` above.
```
Comment (samuellees) [inline tests/moe/test_trtllm_cutlass_fused_moe.py]: what is test_moe_w4a8?

File: tests/moe/test_trtllm_cutlass_fused_moe.py [hunk 7]
```diff
@@ -2400,5 +2404,413 @@ def test_moe_mxfp8_mxfp4_ndim_padding_safety(
...
+        fc2_act_scale,
+        zero_1,
+        zero_2,
+        fc31_alpha,
+        fc2_alpha,
+    )
+
+    routing_weights, selected_experts = compute_routing(router_logits, top_k)
+    flash_output = torch.zeros_like(x)
+    with autotune(True) if use_autotune else nullcontext():
```
Comment (samuellees) [inline tests/moe/test_trtllm_cutlass_fused_moe.py:2784]: @StudyingShao  Could you help check if autotune is used in a correct way, please? Thanks!

File: tests/moe/test_trtllm_cutlass_fused_moe.py (hunk 1 above)
Comment (samuellees) [reply tests/moe/test_trtllm_cutlass_fused_moe.py]: Done. Recover and add extra unit test for this PR

**Comment by samuellees** [comment]
@flashinfer-bot rerun failed

**Review by aleozlx — APPROVED** [review]
lgtm
